internals/plan/plan.go

```diff
@@ -1020,19 +1121,80 @@ func (p *Plan) checkCycles() error {
 }
 
 func ParseLayer(order int, label string, data []byte) (*Layer, error) {
-	layer := Layer{
-		Services:   map[string]*Service{},
-		Checks:     map[string]*Check{},
-		LogTargets: map[string]*LogTarget{},
-	}
-	dec := yaml.NewDecoder(bytes.NewBuffer(data))
-	dec.KnownFields(true)
-	err := dec.Decode(&layer)
+	layer := &Layer{
+		Services:   make(map[string]*Service),
+		Checks:     make(map[string]*Check),
+		LogTargets: make(map[string]*LogTarget),
+		Sections:   make(map[string]LayerSection),
+	}
+
+	// The following manual approach is required because:
+	//
+	// 1. Extended sections are YAML inlined, and also do not have a
+	// concrete type at this level, we cannot simply unmarshal the layer
+	// at once.
+	//
+	// 2. We honor KnownFields = true behaviour for non extended schema
+	// sections, and at the top field level, which includes Section field
+	// names.
+
+	builtinSections := map[string]interface{}{
+		"summary":     &layer.Summary,
+		"description": &layer.Description,
+		"services":    &layer.Services,
+		"checks":      &layer.Checks,
+		"log-targets": &layer.LogTargets,
+	}
+
+	var layerSections map[string]yaml.Node
+	err := yaml.Unmarshal(data, &layerSections)
 	if err != nil {
 		return nil, &FormatError{
 			Message: fmt.Sprintf("cannot parse layer %q: %v", label, err),
 		}
 	}
+
+	for field, section := range layerSections {
+		switch field {
+		case "summary", "description", "services", "checks", "log-targets":
```

Comment:
nit (so we don't have to keep an up-to-date slice of fields twice):
```suggestion
		if _, builtin := builtinSections[field]; builtin {
				// ... case "summary", "description", "services", "checks", "log-targets":
		} else {
				// ... default:
		}
```

Reply:
Fixed this. Great proposal. Resolving.